File: thinking-categorically/state-monad/07-stateful-validation-pattern/stateful-validation-pattern.ts
class State<S, A> {
    constructor(private runState: (state: S) => [A, S]) {}

    run(initialState: S): [A, S] {
        return this.runState(initialState);
    }

    map<B>(f: (a: A) => B): State<S, B> {
        return new State(s => {
            const [a, newS] = this.runState(s);
            return [f(a), newS];
        });
    }

    flatMap<B>(f: (a: A) => State<S, B>): State<S, B> {
        return new State(s => {
            const [a, newS] = this.runState(s);
            return f(a).run(newS);
        });
    }

    then<B>(next: State<S, B>): State<S, B> {
        return this.flatMap(_ => next);
    }

    static pure<S, A>(value: A): State<S, A> {
        return new State(s => [value, s]);
    }

    static get<S>(): State<S, S> {
        return new State(s => [s, s]);
    }

    static put<S>(newState: S): State<S, void> {
        return new State(_ => [undefined, newState]);
    }
}

// Validation state containing errors and context
interface ValidationState {
    errors: string[];
    context: Map<string, string>;
}

// Validate that a field is required (not empty)
function validateRequired(field: string, value: string): State<ValidationState, boolean> {
    return State.get<ValidationState>()
        .flatMap(state => {
            if (!value || value.trim() === '') {
                const newState: ValidationState = {
                    errors: [...state.errors, `${field} is required`],
                    context: new Map(state.context)
                };
                return State.put<ValidationState>(newState).map(() => false);
            } else {
                const newState: ValidationState = {
                    errors: [...state.errors],
                    context: new Map(state.context).set(field, value)
                };
                return State.put<ValidationState>(newState).map(() => true);
            }
        });
}

// Validate email format
function validateEmail(email: string): State<ValidationState, boolean> {
    return State.get<ValidationState>()
        .flatMap(state => {
            const isValid = email.includes('@') && email.includes('.');
            if (!isValid) {
                const newState: ValidationState = {
                    errors: [...state.errors, 'Invalid email format'],
                    context: new Map(state.context)
                };
                return State.put<ValidationState>(newState).map(() => false);
            }
            return State.pure(true);
        });
}

// Validate a complete user
function validateUser(name: string, email: string): State<ValidationState, boolean> {
    return validateRequired("name", name)
        .flatMap(nameValid =>
            validateRequired("email", email)
                .flatMap(emailValid => {
                    if (emailValid) {
                        return validateEmail(email)
                            .map(emailFormatValid => nameValid && emailValid && emailFormatValid);
                    } else {
                        return State.pure(nameValid && emailValid);
                    }
                })
        );
}

// Usage example
const initialState: ValidationState = {
    errors: [],
    context: new Map()
};

const [isValid, finalState] = validateUser("John", "john@example.com").run(initialState);
console.log(`Is valid: ${isValid}`);
console.log(`Errors: ${JSON.stringify(finalState.errors)}`);
console.log(`Context:`, Array.from(finalState.context.entries()));

File: thinking-categorically/state-monad/07-stateful-validation-pattern/stateful-validation-pattern.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Unit type for void-like operations
public struct Unit
{
    public static readonly Unit Default = new Unit();
}

public class State<S, A>
{
    private readonly Func<S, (A Result, S NewState)> computation;

    public State(Func<S, (A, S)> computation)
    {
        this.computation = computation;
    }

    public (A Result, S NewState) Run(S initialState)
    {
        return computation(initialState);
    }

    public State<S, B> Map<B>(Func<A, B> f)
    {
        return new State<S, B>(s =>
        {
            var (result, newState) = computation(s);
            return (f(result), newState);
        });
    }

    public State<S, B> FlatMap<B>(Func<A, State<S, B>> f)
    {
        return new State<S, B>(s =>
        {
            var (result, newState) = computation(s);
            return f(result).Run(newState);
        });
    }

    // LINQ Support
    public State<S, B> SelectMany<B>(Func<A, State<S, B>> f)
    {
        return FlatMap(f);
    }

    public State<S, C> SelectMany<B, C>(Func<A, State<S, B>> f, Func<A, B, C> projection)
    {
        return FlatMap(a => f(a).Map(b => projection(a, b)));
    }

    public State<S, B> Select<B>(Func<A, B> f)
    {
        return Map(f);
    }

    public static State<S, A> Pure(A value)
    {
        return new State<S, A>(s => (value, s));
    }

    public static State<S, S> Get()
    {
        return new State<S, S>(s => (s, s));
    }

    public static State<S, Unit> Put(S newState)
    {
        return new State<S, Unit>(_ => (Unit.Default, newState));
    }
}

// Validation state containing errors and context
public class ValidationState
{
    public List<string> Errors { get; set; }
    public Dictionary<string, string> Context { get; set; }

    public ValidationState(List<string>? errors = null, Dictionary<string, string>? context = null)
    {
        Errors = errors ?? new List<string>();
        Context = context ?? new Dictionary<string, string>();
    }

    public override string ToString()
    {
        return $"ValidationState(Errors: [{string.Join(", ", Errors.Select(e => $"\"{e}\""))}], " +
               $"Context: {{{string.Join(", ", Context.Select(kv => $"{kv.Key}: \"{kv.Value}\""))}}})";
    }
}

public static class Program
{
    // Validate that a field is required (not empty)
    public static State<ValidationState, bool> ValidateRequired(string field, string value)
    {
        return from state in State<ValidationState, ValidationState>.Get()
               let isEmpty = string.IsNullOrWhiteSpace(value)
               let newState = isEmpty
                   ? new ValidationState(
                       state.Errors.Concat(new[] { $"{field} is required" }).ToList(),
                       new Dictionary<string, string>(state.Context))
                   : new ValidationState(
                       new List<string>(state.Errors),
                       new Dictionary<string, string>(state.Context) { [field] = value })
               from _ in State<ValidationState, Unit>.Put(newState)
               select !isEmpty;
    }

    // Validate email format
    public static State<ValidationState, bool> ValidateEmail(string email)
    {
        return from state in State<ValidationState, ValidationState>.Get()
               let isValid = !string.IsNullOrEmpty(email) && email.Contains("@") && email.Contains(".")
               let newState = isValid
                   ? state
                   : new ValidationState(
                       state.Errors.Concat(new[] { "Invalid email format" }).ToList(),
                       new Dictionary<string, string>(state.Context))
               from _ in State<ValidationState, Unit>.Put(newState)
               select isValid;
    }

    // Validate a complete user
    public static State<ValidationState, bool> ValidateUser(string name, string email)
    {
        return from nameValid in ValidateRequired("name", name)
               from emailValid in ValidateRequired("email", email)
               from emailFormatValid in emailValid ? ValidateEmail(email) : State<ValidationState, bool>.Pure(false)
               select nameValid && emailValid && emailFormatValid;
    }

    // Usage example
    public static void Main()
    {
        var initialState = new ValidationState();
        var (isValid, finalState) = ValidateUser("John", "john@example.com").Run(initialState);

        Console.WriteLine($"Is valid: {isValid}");
        Console.WriteLine($"Final state: {finalState}");
    }
}
